jupyter_scheduler/scheduler.py

```diff
@@ -247,16 +257,84 @@ def stop_job(self, job_id):
-        pass
+            job_definition_id = job_definition.job_definition_id
+
+        return job_definition_id
+
+    def update_job_definition(self, model: UpdateJobDefinition):
+        with self.db_session() as session:
+            session.query(JobDefinition).filter(
+                JobDefinition.job_definition_id == model.job_definition_id
+            ).update(model.dict(exclude_none=True))
+            session.commit()
 
     def delete_job_definition(self, job_definition_id: str):
-        pass
+        with self.db_session() as session:
+            jobs = session.query(Job).filter(Job.job_definition_id == job_definition_id)
+            for job in jobs:
+                self.delete_job(job.job_id)
+            session.query(JobDefinition).filter(
+                JobDefinition.job_definition_id == job_definition_id
+            ).delete()
+            session.commit()
 
-    def list_job_definitions(self, query: ListJobDefinitionsQuery) -> List[DescribeJobDefinition]:
-        pass
+    def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinitionsResponse:
+        with self.db_session() as session:
+            definitions = session.query(JobDefinition)
+
+            if query.create_time:
+                definitions = definitions.filter(JobDefinition.create_time >= query.create_time)
+            if query.name:
+                definitions = definitions.filter(JobDefinition.name.like(f"{query.name}%"))
+            if query.tags:
+                definitions = definitions.filter(
+                    and_(JobDefinition.tags.contains(tag) for tag in query.tags)
+                )
+
+            total = definitions.count()
+
+            if query.sort_by:
+                for sort_field in query.sort_by:
+                    direction = desc if sort_field.direction == SortDirection.desc else asc
+                    definitions = definitions.order_by(
+                        direction(getattr(JobDefinition, sort_field.name))
+                    )
+            next_token = int(query.next_token) if query.next_token else 0
+            definitions = definitions.limit(query.max_items).offset(next_token)
+
+            definitions = definitions.all()
+
+        next_token = next_token + len(definitions)
+        if next_token >= total:
+            next_token = None
+
+        list_response = ListJobDefinitionsResponse(
+            job_definitions=[
+                DescribeJobDefinition.from_orm(definition) for definition in definitions or []
+            ],
+            next_token=next_token,
+            total_count=total,
+        )
+
+        return list_response
 
     def pause_jobs(self, job_definition_id: str):
-        pass
+        with self.db_session() as session:
+            session.query(JobDefinition).filter(
+                JobDefinition.job_definition_id == job_definition_id
+            ).update(active=False)
+            session.commit()
```

Comment:
1) It looks like this isn't actually pausing job execution if the job's already running. Is this expected behavior?

2) `session.query(JobDefinition)` seems to be querying for all the job definitions rather than the all the jobs. Are you sure this shouldn't be `session.query(Job)` instead?

Reply:
The intent here was to stop all future job executions, so the code here is correct.

Reply:
Yeah, I don't think that the pause action makes sense for actively running jobs.